crates/sui-rpc-benchmark/src/json_rpc/runner.rs

```diff
@@ -54,51 +76,179 @@ impl JsonRpcStats {
+                            "params not found or not an array for method: {}",
+                            request_line.method
+                        )
+                    })?;
+
+                if let Some(cursor_idx) =
+                    PaginationCursorState::get_method_cursor_index(&request_line.method)
+                {
+                    let method_key =
+                        PaginationCursorState::get_method_key(&request_line.method, &params)?;
+                    PaginationCursorState::update_params_cursor(
+                        &mut request_line.body_json,
+                        cursor_idx,
+                        pagination_state.get(&method_key).as_ref(),
+                        &request_line.method,
+                    )?;
+                }
+
+                let now = Instant::now();
+                let res = timeout(
+                    Duration::from_secs(10),
+                    client.post(&endpoint).json(&request_line.body_json).send(),
+                )
+                .await;
+                let elapsed_ms = now.elapsed().as_millis() as f64;
+
+                // update pagination cursor if the request is successful.
+                let mut is_error = true;
+                if let Ok(Ok(resp)) = res {
+                    if resp.status().is_success() {
+                        #[derive(Deserialize)]
+                        struct Body {
+                            result: Result,
+                        }
+                        #[derive(Deserialize)]
+                        #[serde(rename_all = "camelCase")]
+                        struct Result {
+                            has_next_page: bool,
+                            next_cursor: Option<Value>,
+                        }
+
+                        if let Ok(Body { result }) = resp.json::<Body>().await {
+                            let method_key = PaginationCursorState::get_method_key(
+                                &request_line.method,
+                                &params,
+                            )?;
+                            pagination_state.update(
+                                method_key,
+                                if result.has_next_page {
+                                    result.next_cursor
+                                } else {
+                                    None
+                                },
+                            );
+                            is_error = false;
+                        }
+                    }
+                }
 
-    timeout(
-        config.duration,
-        stream.try_for_each_spawned(concurrency, |fut| fut),
-    )
-    .await
-    .unwrap_or(Ok(()))?;
+                // Record stats after all async operations to avoid error of sending future between threads
+                let mut stats = task_stats
+                    .lock()
+                    .map_err(|e| anyhow::anyhow!("Failed to acquire stats lock: {}", e))?;
```

Comment:
nit: just `unwrap` / `expect` here -- we should panic if the lock has been poisoned.

Reply:
yes, sorry missed this one, good catch!